server/routerlicious/packages/services-telemetry/src/resources.ts

```diff
@@ -118,19 +118,21 @@ export interface ILumberjackSchemaValidationResult {
 
 // Helper method to assist with handling Lumberjack/Lumber errors depending on the context.
 export function handleError(eventName: LumberEventName, errMsg: string, engineList: ILumberjackEngine[]) {
-    const err = new Error(errMsg);
-    // If there is no LumberjackEngine specified, making the list empty,
-    // we log the error to the console as a last resort, so the information can
-    // be found in raw logs.
-    if (engineList.length === 0) {
-        console.error(serializeError(err));
-    } else {
-        // Otherwise, we log the error through the current LumberjackEngines.
-        const errLumber = new Lumber<LumberEventName>(
-            eventName,
-            LumberType.Metric,
-            engineList);
-        errLumber.error(errMsg, err);
+    if (process.env.IS_SERVER) {
```

Comment:
process is not isomorphic. it will not work in the browser. so this solution will not work.

Reply:
Lumberjack is not supposed to be isomorphic. It is not included as part of serices-client, which is the package that, within /server, should provide isomorphic code.

Reply:
local-server, and the routerlicious driver needs to remain isomorphic. the original problem is that clients, which only use local-server and/or the driver are seeing: https://github.com/microsoft/FluidFramework/blob/f3e44371c498c2d89951de3167e71946df3c04f9/server/routerlicious/packages/services-telemetry/src/lumberjack.ts#L226

i would think the fix to that would also need to isomorphic.

Reply:
Routerlicious driver does not use Lumberjack. Now, as for local-server, let me try to clarify something as I'm curious: [local-server depends on lambdas and services-core](https://github.com/microsoft/FluidFramework/blob/965b46082a2626542364b4587337f2c6b16e483d/server/routerlicious/packages/local-server/package.json#L61-L69). l[ambdas also depends on lambdas-driver](https://github.com/microsoft/FluidFramework/blob/965b46082a2626542364b4587337f2c6b16e483d/server/routerlicious/packages/lambdas/package.json#L59). Does that mean that all those packages need to be isomorphic? Do we have documentation anywhere explaining what packages need to be isomorphic, and which don't?

Reply:
Also, I think process not being isomorphic is a problem we have already solved in webpack. I'm talking about this part here:
https://github.com/microsoft/FluidFramework/blob/09df4d28a5dab14d3b61ff2bb68fc1d2de1b42b0/packages/tools/webpack-fluid-loader/webpack.config.js#L38-L42

Maybe adding something like the below could fix this issue for us?
```ts
new webpack.DefinePlugin({ 
       process: { 
           env: { 
                NODE_ENV: JSON.stringify(mode) ,
                IS_SERVER: false
           }
        }, 
       global : {} 
     }), 
```

Reply:
No documentation as far as i know. There is a test, but things have slipped through the cracks in the past:
https://github.com/microsoft/FluidFramework/blob/main/server/routerlicious/packages/local-server/src/test/isomorphic.spec.ts

Moving to webpack 5 #9568 would help, as it fails on missing polyfils more reliably. 

The problem with any webpack config is we don't control it. we have some examples that webpack, and you could update those, but that will have no impact on downstream consumers, and they will all have their own custom config.

Reply:
local server is used in a ton of places. partners rely on it heavily for testing and demoing, even our docs depend on it:
https://fluidframework.com/docs/start/tutorial/

Reply:
I see. Thank you for sharing all that information. Do you have other suggestions/ideas that could help us achieve the same goal while still being isomorphic? If not, I think the only immediate option for me is to remove the Lumberjack logs completely, which I was trying to avoid.

Reply:
i don't really understand the error, or it's use. why not just throw, and wherever it's not being set up just pass a dummy logger

Reply:
Throwing did not sound right because we don't want make the service fail due to logging setup issues. As for setting up Lumberjack properly, I already did all the setup, including in local-server (using a dummy logger). The problem is that due to package version mismatch, local-server might have it's own copy of server-telemetry. It sets up Lumberjack, but later, when one of its dependencies, say services-core, used Lumberjack, it ends up being a different instance, since services-core has its own inner copy of services-telemetry. That is one behavior we have observed only when services-telemetry ends up being an indirect dependency of the client of FFX unit tests. We discussed about that problem in detail here: https://github.com/microsoft/FluidFramework/issues/9370

Let me think if there is a different way to do it, and if not, I will just remove all the logs to unblock our partners.

Reply:
i see. that is a frequent problem with the singleton pattern in js. at this point are the errors even useful if the package is setup everywhere it needs to be?

Reply:
That's correct @anthony-murphy. The errors have been helpful on the server side but I did not anticipate they would cause so much pain elsewhere, unfortunately. Sorry about that. Last try before completely disabling all this console logging. One idea @znewton suggested to me (thanks @znewton !) was also checking if `window` is `undefined` to detect whether we are running on the browser or not. So we could have something like:
```
if (typeof window === "undefined" && process.env.NODE_ENV) {
```
Do you think that could be an acceptable solution? The first condition would make sure we are running on Node.js, and the second condition would make sure we are not running UTs (assuming UTs run without setting `NODE_ENV`. I could also use the new `IS_SERVER` variable I just added).

Reply:
I don't know what NODE_ENV is. if that is something you all control then it is probably fine. I think you should also be more defensive around the process check `process?.env?.NODE_ENV` incase some environment doesn't have window or process, like service workers, react native, or some environment i've never heard of

Reply:
in general you can't really control what environment variables exit, both NODE_ENV and IS_SERVER seem a little general to me. I would go with something more specific like IS_FLUID_SERVER

Reply:
Thanks for the feedback @anthony-murphy. I can certainly be more defensive when doing the process check. As for `NODE_ENV`, it's not something we control entirely, as some non-server files in this repo seem to set that. (examples [here](https://github.com/microsoft/FluidFramework/blob/9c767dc4834c0bd2894fe21a0ce491951854c0c9/experimental/dds/tree/.env-cmdrc.json), [here](https://github.com/microsoft/FluidFramework/blob/9c767dc4834c0bd2894fe21a0ce491951854c0c9/experimental/dds/tree-graphql/.env-cmdrc.json), [here](https://github.com/microsoft/FluidFramework/blob/09df4d28a5dab14d3b61ff2bb68fc1d2de1b42b0/packages/tools/webpack-fluid-loader/webpack.config.js) and [here](https://github.com/microsoft/FluidFramework/blob/602cf06fb312b9ca1a6064724d56af7d63e0b9e3/docs/themes/thxvscode/assets/js/webpack.config.js)). That's why I added `IS_SERVER`. Even if it can be redundant to `NODE_ENV` in some cases, we can use it exclusively here. I think you've got a good point to rename it to `IS_FLUID_SERVER`, which is more specific. I also didn't like `IS_SERVER` much, it sounded a bit weird.

I will make the changes to look like this:

```ts
if (typeof window === "undefined" && process?.env?.IS_FLUID_SERVER) {
```

I can quickly test that in r11s and unit test environments (in r11s, which I'm guessing would be similar to FFX UTs). Do you have recommendations on how else to test this, for example in the browser? Or does that require a pre-release of the server packages first?

Reply:
[This SO answer](https://stackoverflow.com/a/31090240/551579) is quite detailed. The utility functions described there might be useful to add for both clarity and consistency in how we do environment detection. There's also [an npm package](https://www.npmjs.com/package/browser-or-node) but not worth adding an external dependency IMO. But I do like a wrapper function with an easy-to-understand name.

Reply:
For posterity, the NODE_ENV is (was?) a commonly used convention and shows up in a lot of getting started guides. I think it was popularized by express, where it's used to _express_ (see what I did there? :smile:) production vs. development mode. But ultimately it's just an environment variable like any other, which express and maybe other libs have unified around.

The examples in the repo where we use environment variables in code that's bundled for the browser are using [Webpack plugins](https://webpack.js.org/plugins/define-plugin) to replace `'process.env.NODE_ENV': process.env.NODE_ENV`. That is, it replaces the raw string `process.env.NODE_ENV` with the results of calling `process.env.NODE_ENV`. So in a packed bundle you'll see code like `if('production' === 'production')` because it's straight-up string replacement.

Reply:
Thank you for all the information @tylerbutler. It was a good suggestion to have helpers to make it explicit we are trying to detect the environment. I pushed new changes that include that. Can you and @anthony-murphy take a look and tell me what you think?

Also, what would be the best way to test this in the context of FFX tests/browser?

Reply:
i think this will work fine as a mitigation. 

in terms of testing. i would get this into main, and then bump the server version in client, and run some of our browser examples, and unit tests. i don't expect issues.

we then need to figure out if this needs back porting, probably to 0.58 and/or 0.59. once there we can release and have them test it. i don't think we have a good system for testing pre-release

Reply:
something i might consider long term is moving away from the singleton pattern here. that should allow fail fast which would be better, but this is low priority.

Reply:
Thanks @anthony-murphy! Yes, I already have something in mind, at least for all the packages we talked above that need to be isomorphic. We can talk more about that in the near future